src/core/core-varlink.c

```diff
@@ -613,20 +637,35 @@ static int vl_reply(sd_varlink *link, sd_json_variant *parameters, const char *e
 }
 
 static int manager_varlink_init_user(Manager *m) {
+        char address[STRLEN("/run/user//systemd/io.systemd.Manager") + DECIMAL_STR_MAX(uid_t)];
         _cleanup_(sd_varlink_close_unrefp) sd_varlink *link = NULL;
         int r;
 
         assert(m);
 
-        if (m->managed_oom_varlink)
-                return 1;
-
         if (!MANAGER_IS_USER(m))
                 return 0;
 
         if (MANAGER_IS_TEST_RUN(m))
                 return 0;
 
+        r = manager_setup_varlink_server(m);
+        if (r < 0)
+                return log_error_errno(r, "Failed to set up varlink server: %m");
+        bool fresh = r > 0;
+
+        xsprintf(address, "/run/user/" UID_FMT "/systemd/io.systemd.Manager", geteuid());
+
+        /* We might have got sockets through deserialization. Do not bind to them twice. */
+        if (fresh || !varlink_server_contains_socket(m->varlink_server, address)) {
+                r = sd_varlink_server_listen_address(m->varlink_server, address, 0666);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to bind to varlink socket '%s': %m", address);
+        }
+
+        if (m->managed_oom_varlink)
```

Comment:
Hmm, I think the part handling managed oom connection is worth splitting into another function, and call that in `manager_varlink_send_managed_oom_update`. Otherwise, each `manager_varlink_send_managed_oom_update` call would needlessly re-run this bit of logic.